server/controllers/roomController.ts

```diff
@@ -0,0 +1,122 @@
+import { Room, RoomClass } from "../db/room";
+import { Player, PlayerClass } from "../db/player";
+import { DocumentType } from "@typegoose/typegoose";
+import { Card } from "../db/card";
+
+function shuffle<T>(array: T[]): T[] {
+  var currentIndex = array.length,
+    temporaryValue,
+    randomIndex;
+
+  // While there remain elements to shuffle...
+  while (0 !== currentIndex) {
+    // Pick a remaining element...
+    randomIndex = Math.floor(Math.random() * currentIndex);
+    currentIndex -= 1;
+
+    // And swap it with the current element.
+    temporaryValue = array[currentIndex];
+    array[currentIndex] = array[randomIndex];
+    array[randomIndex] = temporaryValue;
+  }
+
+  return array;
+}
+
+export class RoomController {
+  static HAND_SIZE = 12;
+  /**
+   * Removes a {@link Player} from any {@link Room}s they are part of, given only their connectionId
+   * @param connectionId The connectionId of the {@link Player} to remove.
+   */
+  async removePlayer(
+    connectionId: string
+  ): Promise<DocumentType<RoomClass>[] | undefined> {
+    const player = await Player.findOne({ connectionId });
+    if (!player) {
+      console.error(
+        `No player with connection ID ${connectionId} was found to remove from rooms.`
+      );
+      return;
+    }
+    // Get IDs of the rooms affected
+    const affectedRooms = await Room.find({ players: { $all: [player._id] } });
+    const affectedRoomIds = affectedRooms.map((room) => room._id);
+
+    await Room.updateMany(
+      { players: { $all: [player._id] } },
+      { $pull: { players: player._id } }
+    );
+
+    await Player.deleteMany({ connectionId }).exec();
+
+    return await Room.find({ _id: { $in: affectedRoomIds } }).populate(
+      "players"
+    );
+  }
+  /**
+   * Creates and returns a new {@link Room}.
+   * @param open Whether or not the {@link Room} should be accessible by the public
+   */
+  public async createRoom(open: boolean = false) {
+    // Just in case there are collisions when generating the code, retry until no collisions.
+    while (true) {
+      try {
+        return await Room.create({ open });
+      } catch (err) {
+        console.error(err);
```

Comment:
Break the while loop here if there is an error? What if there is some issue with the DB and this just hammers it because infinite retries.

Reply:
Good point. I can make it so that the error will actually throw if it's not a unique key error.